tools/src/main/scala/caliban/tools/SchemaWriter.scala

```diff
@@ -152,29 +153,60 @@ object SchemaWriter {
           .getOrElse(Type.innerType(field.ofType), List.empty)
           .exists(t => hasFieldWithDirective(t, directive))
 
+      val field = resolveTypesafeFieldDef(inputField).getOrElse(inputField)
+
       val fieldIsEffectWrapped               = field.directives.exists(_.name == LazyDirective)
-      val fieldTypeIsEffectTypeParameterized = isEffectTypeAbstract && containsNestedDirective(LazyDirective)
+      val fieldTypeIsEffectTypeParameterized = isEffectTypeAbstract && containsNestedDirective(field, LazyDirective)
       val fieldType                          = (fieldIsEffectWrapped, fieldTypeIsEffectTypeParameterized) match {
         case (true, true)   => s"$effect[${writeParameterizedType(field.ofType)}]"
         case (true, false)  => s"$effect[${writeType(field.ofType)}]"
         case (false, true)  => writeParameterizedType(field.ofType)
         case (false, false) => writeType(field.ofType)
       }
+
+      val GQLTypesafeDirective = writeGQLTypesafeDirective(field)
+
       if (field.args.nonEmpty) {
-        s"${writeDescription(field.description)}${if (isMethod) "def " else ""}${safeName(field.name)} : ${argsName(field, of)} => $fieldType"
+        s"""$GQLTypesafeDirective${writeDescription(field.description)}${if (isMethod) "def " else ""}${safeName(
+          field.name
+        )} : ${argsName(field, of)} => $fieldType"""
       } else {
-        s"""${writeDescription(field.description)}${if (isMethod) "def " else ""}${safeName(
+        s"""$GQLTypesafeDirective${writeDescription(field.description)}${if (isMethod) "def " else ""}${safeName(
           field.name
         )} : $fieldType"""
       }
     }
 
-    def writeInputValue(value: InputValueDefinition): String =
-      s"""${writeDescription(value.description)}${safeName(value.name)} : ${writeType(value.ofType)}"""
+    def writeGQLTypesafeDirective(field: FieldDefinition) =
+      if (field.directives.exists(_.name == TypesafeDirective)) {
```

Comment:
nitpick: rather than `exists` + `filter` + `head`, you could use `find` and then `fold` on the Option. Same below.

Reply:
Thanks I even found some further optimization cause of this :)